docs/microbenchmark-design-guidelines.md

```diff
@@ -0,0 +1,557 @@
+
+`ValuesGenerator` class from `BenchmarkDotNet.Extensions` library exposes following methods that guarantee to always produce the same data:
+
+```cs
+T GetNonDefaultValue<T>();
+T[] ArrayOfUniqueValues<T>(int count);
+T[] Array<T>(int count);
+Dictionary<TKey, TValue> Dictionary<TKey, TValue>(int count)
+```
+
+As of today, the `T` can be: `byte`, `char`, `int`, `double`, `bool` and `string`. Extending `T` to more types is very welcomed!
+
+**Note:** `ValuesGenerator` is simply always creating a new instance of `Random` with a constant seed. It's a crucial component and its correctness is verified using  [Unit Tests](https://github.com/dotnet/performance/blob/master/src/tests/harness/BenchmarkDotNet.Extensions.Tests/UniqueValuesGeneratorTests.cs).
+
+**Note:** Please don't use `Random` directly in the benchmarks, do use `ValuesGenerator` and extend it with missing features when needed.
+
+### BenchmarkDotNet
+
+#### Arguments
+
+`[Arguments]` is xUnit's `[InlineData]` counterpart.
+
+```cs
+[Benchmark]
+[Arguments(100)]
+public void Method(int length)
+```
+
+**Note:** BenchmarkDotNet supports passing arguments by `ref` to the benchmarked method.
+
+#### Params
+
+If you need to use the argument in the setup method, then instead of using `Arguments` you should use `Params`. A `[Params]` field|property can be accessed during the setup/cleanup ([docs](https://benchmarkdotnet.org/Advanced/Params.htm)). Example:
+
+```cs
+int[] _array;
+
+[Params(
+    100, // comment explaining why it's needed to have this test case
+    1_000, // comment explaining why it's needed to have this test case
+    10_000)] // comment explaining why it's needed to have this test case
+public int Size { get; set; }
+
+[GlobalSetup]
+public void PrepareArray() => _array = new int[Size];
+```
+
+However, **`[Params]` are applied to all the benchmarks in given class**. So if few benchmarks require different `[Params]` values they have to be split into separate types (current BenchmarkDotNet implementation limitation).
+
+#### ArgumentsSource
+
+`[ArgumentsSource]` is xUnit's `[MemberData]` counterpart.
+
+```cs
+// for single argument it's an IEnumerable of objects (object), not IEnumerable of arrays of objects (object[])
+public IEnumerable<object> Cultures()
+{
+    yield return new CultureInfo("fr");
+    yield return new CultureInfo("da");
+    yield return new CultureInfo("ja");
+    yield return new CultureInfo("");
+}
+
+[Benchmark]
+[ArgumentsSource(nameof(Cultures))]
+public DateTime Parse(CultureInfo cultureInfo)
+    => DateTime.Parse("10/10/2010 12:00:00 AM", cultureInfo);
+```
+
+**Note:** If you need to use the argument in the setup method, then instead of using `ArgumentsSource` you should use `ParamsSource`.
+
+**Note:** the time spent for initializing an argument is not included in the time reported by the harness. It means that you can have a custom complex type with initialization logic in it's ctor and setup the benchmark in that way:
+
```

Comment:
It might be worth noting here that `ArgumentsSource` methods are only called once per execution of the benchmark, and the cached results are used per invocation. This is something that confused me for a while.

Reply:
Another good point! It makes me wonder what are the other things that are obvious to me but surprising to the end user ;)

Reply:
Ok, let's see what the users are going to say https://twitter.com/SitnikAdam/status/1148555464394911744